File: tests/csharp/test_envelope_sdk.cs
/**
 * Test for envelope SDK interface with field_order discriminator.
 *
 * Regression test for the enum naming bug in SdkInterface.cs where the
 * discriminator enum was referenced without the package prefix (e.g.
 * "RawDataEnvelopePayloadField") instead of the correct full name
 * (e.g. "RawDataEnvelopePayloadField"), causing CS0266
 * implicit conversion errors when building the SDK interface.
 *
 * This test verifies:
 * 1. The discriminator enum is accessible with the correct package-prefixed name.
 * 2. Enum values follow the field_order (1-based) convention.
 * 3. RawDataEnvelope can be round-trip serialized with each payload variant.
 * 4. The discriminator field is correctly set/read after deserialization.
 */

using System;
using StructFrame.EnvelopeTest;

public class TestEnvelopeSdk
{
  private static int _passed = 0;
  private static int _failed = 0;

  private static void Expect(bool condition, string testName)
  {
    if (condition)
    {
      Console.WriteLine($"  PASS  {testName}");
      _passed++;
    }
    else
    {
      Console.Error.WriteLine($"  FAIL  {testName}");
      _failed++;
    }
  }

  /**
   * Verify discriminator enum values follow field_order (1-based).
   * Before the fix this file would not compile due to the wrong enum type name
   * being referenced in SdkInterface.cs.
   */
  private static void TestDiscriminatorEnumValues()
  {
    // RawDataEnvelopePayloadField is the correct full name.
    // If SdkInterface.cs used the wrong name, the generated library would
    // fail to build and this test file would never reach execution.
    Expect((byte)RawDataEnvelopePayloadField.None == 0,
           "Discriminator None == 0");
    Expect((byte)RawDataEnvelopePayloadField.Sample == 1,
           "Discriminator Sample == 1 (first field in oneof)");
    Expect((byte)RawDataEnvelopePayloadField.Config == 2,
           "Discriminator Config == 2 (second field in oneof)");
  }

  /**
   * Round-trip test: serialize a RawDataEnvelope with a RawSamplePayload,
   * deserialize it, and confirm the discriminator and payload fields survive.
   */
  private static void TestRoundTripSamplePayload()
  {
    var sample = new RawSamplePayload
    {
      Channel = 3,
      Value = 1.23f,
      Flags = 0xAB
    };

    var envelope = new RawDataEnvelope
    {
      Priority = 5,
      TimestampUs = 1000000,
      PayloadDiscriminator = RawDataEnvelopePayloadField.Sample,
      Sample = sample
    };

    byte[] bytes = envelope.Serialize();
    var decoded = RawDataEnvelope.Deserialize(bytes);

    Expect(decoded.Priority == 5, "RoundTrip Sample: Priority");
    Expect(decoded.TimestampUs == 1000000, "RoundTrip Sample: TimestampUs");
    Expect(decoded.PayloadDiscriminator == RawDataEnvelopePayloadField.Sample,
           "RoundTrip Sample: Discriminator == Sample");
    Expect(decoded.Sample != null, "RoundTrip Sample: Sample field not null");
    if (decoded.Sample != null)
    {
      Expect(decoded.Sample.Channel == 3, "RoundTrip Sample: Channel");
      Expect(Math.Abs(decoded.Sample.Value - 1.23f) < 1e-5f, "RoundTrip Sample: Value");
      Expect(decoded.Sample.Flags == 0xAB, "RoundTrip Sample: Flags");
    }
  }

  /**
   * Round-trip test: serialize a RawDataEnvelope with a RawConfigPayload.
   */
  private static void TestRoundTripConfigPayload()
  {
    var config = new RawConfigPayload
    {
      SettingId = 7,
      SettingValue = 4096
    };

    var envelope = new RawDataEnvelope
    {
      Priority = 1,
      TimestampUs = 500,
      PayloadDiscriminator = RawDataEnvelopePayloadField.Config,
      Config = config
    };

    byte[] bytes = envelope.Serialize();
    var decoded = RawDataEnvelope.Deserialize(bytes);

    Expect(decoded.Priority == 1, "RoundTrip Config: Priority");
    Expect(decoded.TimestampUs == 500, "RoundTrip Config: TimestampUs");
    Expect(decoded.PayloadDiscriminator == RawDataEnvelopePayloadField.Config,
           "RoundTrip Config: Discriminator == Config");
    Expect(decoded.Config != null, "RoundTrip Config: Config field not null");
    if (decoded.Config != null)
    {
      Expect(decoded.Config.SettingId == 7, "RoundTrip Config: SettingId");
      Expect(decoded.Config.SettingValue == 4096, "RoundTrip Config: SettingValue");
    }
  }

  public static int Main(string[] args)
  {
    Console.WriteLine("=== EnvelopeSdk field_order discriminator tests ===");

    TestDiscriminatorEnumValues();
    TestRoundTripSamplePayload();
    TestRoundTripConfigPayload();

    Console.WriteLine();
    Console.WriteLine($"Summary: {_passed} passed, {_failed} failed");

    return _failed > 0 ? 1 : 0;
  }
}

File: tests/cpp/test_envelope_sdk.cpp
/**
 * Test for envelope SDK interface: msgid-discriminator (CommandEnvelope) and
 * field_order-discriminator (RawDataEnvelope) round-trips.
 *
 * Mirrors tests/rust/src/main.rs::run_envelope_sdk_tests().
 */

#include <cstdio>
#include <cstring>
#include <cmath>
#include <vector>

#include "frame_profiles.hpp"
#include "envelope_test.structframe.hpp"

using namespace structframe;
using namespace structframe::envelope_test;

static int g_passed = 0;
static int g_failed = 0;

static void expect(bool condition, const char* name) {
  if (condition) {
    printf("  PASS  %s\n", name);
    g_passed++;
  } else {
    printf("  FAIL  %s\n", name);
    g_failed++;
  }
}

int main() {
  printf("=== C++ EnvelopeSdk tests ===\n");

  // --- CommandEnvelope with msgid discriminator ---
  ADCCommand adc{};
  adc.channel = 2;
  adc.sample_rate = 1000;
  adc.enable = true;
  CommandEnvelope env = CommandEnvelope::wrap(42, 5, true, adc);

  expect(env.command_discriminator == ADCCommand::MSG_ID,
         "CommandEnvelope: discriminator == ADCCommand::MSG_ID");

  // Encode with ProfileStandard (MSG_ID=200 fits in u8)
  std::vector<uint8_t> buffer(2048);
  BufferWriter<ProfileStandardConfig> writer(buffer.data(), buffer.size());
  size_t written = writer.write(env);
  expect(written > 0, "CommandEnvelope: write returned > 0");

  AccumulatingReader<ProfileStandardConfig, 4096, decltype(&get_message_info)> reader(get_message_info);
  reader.add_data(buffer.data(), written);
  auto frame = reader.next();
  expect(frame.valid, "CommandEnvelope: frame valid");
  if (frame.valid) {
    expect(frame.msg_id == CommandEnvelope::MSG_ID, "CommandEnvelope: msg_id matches");
    CommandEnvelope decoded;
    size_t n = decoded.deserialize(frame);
    expect(n > 0, "CommandEnvelope: deserialize succeeded");
    expect(decoded.sequence_number == 42, "CommandEnvelope: sequence_number round-trips");
    expect(decoded.priority == 5, "CommandEnvelope: priority round-trips");
    expect(decoded.run_immediately, "CommandEnvelope: run_immediately round-trips");
    expect(decoded.command_discriminator == ADCCommand::MSG_ID,
           "CommandEnvelope: discriminator round-trips");
    expect(decoded.command.adc.channel == 2, "CommandEnvelope: adc.channel round-trips");
    expect(decoded.command.adc.sample_rate == 1000, "CommandEnvelope: adc.sample_rate round-trips");
    expect(decoded.command.adc.enable, "CommandEnvelope: adc.enable round-trips");
  }

  // --- RawDataEnvelope with field_order discriminator ---
  RawSamplePayload sample{};
  sample.channel = 7;
  sample.value = 2.718f;
  sample.flags = 0xBE;
  RawDataEnvelope raw_env = RawDataEnvelope::wrap(3, 999000, sample);

  // field_order discriminator: sample is 1st field -> discriminator == 1
  expect(static_cast<uint8_t>(raw_env.payload_discriminator) == 1,
         "RawDataEnvelope: field_order discriminator == 1 for first variant");

  std::vector<uint8_t> raw_buffer(2048);
  BufferWriter<ProfileBulkConfig> raw_writer(raw_buffer.data(), raw_buffer.size());
  size_t raw_written = raw_writer.write(raw_env);
  expect(raw_written > 0, "RawDataEnvelope: write returned > 0");

  AccumulatingReader<ProfileBulkConfig, 4096, decltype(&get_message_info)> raw_reader(get_message_info);
  raw_reader.add_data(raw_buffer.data(), raw_written);
  auto raw_frame = raw_reader.next();
  expect(raw_frame.valid, "RawDataEnvelope: frame valid");
  if (raw_frame.valid) {
    RawDataEnvelope raw_decoded;
    size_t n = raw_decoded.deserialize(raw_frame);
    expect(n > 0, "RawDataEnvelope: deserialize succeeded");
    expect(raw_decoded.priority == 3, "RawDataEnvelope: priority round-trips");
    expect(raw_decoded.timestamp_us == 999000, "RawDataEnvelope: timestamp_us round-trips");
    expect(static_cast<uint8_t>(raw_decoded.payload_discriminator) == 1,
           "RawDataEnvelope: discriminator round-trips == 1");
    expect(raw_decoded.payload.sample.channel == 7, "RawDataEnvelope: sample.channel round-trips");
    expect(std::abs(raw_decoded.payload.sample.value - 2.718f) < 1e-4f,
           "RawDataEnvelope: sample.value round-trips");
    expect(raw_decoded.payload.sample.flags == 0xBE, "RawDataEnvelope: sample.flags round-trips");
  }

  printf("\nSummary: %d passed, %d failed\n", g_passed, g_failed);
  return g_failed > 0 ? 1 : 0;
}
